client/my-sites/email/email-providers-stacked-comparison/provider-cards/utils.tsx

```diff
@@ -0,0 +1,135 @@
+import formatCurrency from '@automattic/format-currency';
+import {
+	RequestCartProduct,
+	ResponseCart,
+	ShoppingCartManagerActions,
+} from '@automattic/shopping-cart';
+import { translate } from 'i18n-calypso';
+import page from 'page';
+import React, { FunctionComponent, ReactElement } from 'react';
+import PromoCardPrice from 'calypso/components/promo-section/promo-card/price';
+import { recordTracksEvent } from 'calypso/lib/analytics/tracks';
+import { fillInSingleCartItemAttributes } from 'calypso/lib/cart-values';
+import { IncompleteRequestCartProduct } from 'calypso/lib/cart-values/cart-items';
+
+export enum IntervalLength {
+	ANNUALLY = 'annually',
+	MONTHLY = 'monthly',
+}
+
+export const addToCartAndCheckout = (
+	shoppingCartManager: ShoppingCartManagerActions,
+	cartItem: RequestCartProduct | IncompleteRequestCartProduct,
+	productList: Record< string, { product_id: number } >,
+	setAddingToCart: ( addingToCart: boolean ) => void,
+	selectedSite: string
+): void => {
+	setAddingToCart( true );
+
+	shoppingCartManager
+		.addProductsToCart( [ fillInSingleCartItemAttributes( cartItem, productList ) ] )
+		.then( ( response: ResponseCart ) => {
+			setAddingToCart( false );
+			const errors = response?.messages?.errors;
+			if ( errors && errors.length ) {
+				// Stay on the page to show the relevant error(s)
+				return;
+			}
+
+			page( `/checkout/${ selectedSite }` );
+		} )
+		.then( () => {
+			setAddingToCart( false );
+		} );
+};
+
+export const recordTracksEventAddToCartClick = (
+	comparisonContext: string,
+	validatedMailboxUuids: string[],
+	mailboxesAreValid: boolean,
+	provider: string,
+	source: string,
+	userCanAddEmail: boolean,
+	userCannotAddEmailReason: any
```

Comment:
From the code below it looks like you already know this type. Something like `undefined | {code: string}`

Reply:
Thank you for chiming in, that's super useful to me. I'll update it.